src/utils/actions.js

```diff
@@ -0,0 +1,254 @@
+//  contentBundle: just the content section of the report bundle
+//  next: place to put arbitrary objects specifically for the next step?
+//  prev: where arbitrary objects from the previous set should go?
+//  flags: various boolean or enum flags about the status of the report. is it encrypted y/n, is it identifiable or de/pseudo/anonymized, etc
+//  cancelToken: indicates the job should be cancelled (from a process external to the job itself -- this tells the job to cease processing)
+//  exitStatus: if set, terminates the job and marks it with the given status (for example, if a given patient doesn't meet reporting criteria, don't keep reporting)
+
+// we assume the context updates and is passed to functions
+// in turn.  I'm also assuming the context is unique
+// to the current action, since we'd be cycling through actions
+// anyway.
+const Fhir = require('fhir').Fhir;
+const FhirPath = require('fhir').FhirPath;
+const fhir = new Fhir();
+const baseIgActions = {
+  'check-trigger-codes': context => {
+    const action = context.action;
+    const resources = context.records;
+
+    const boolMap = action.condition.map(condition => {
+      if (condition.kind === 'applicability' && condition.expression) {
+        const expression = condition.expression;
+        // check the records for a trigger code
+        return resources.some(resource => {
+          //might have to check that this is non-empty here
+          return evaluateExpression(expression, resource);
+        });
+      } else {
+        // default true for non-applicability conditions
+        return true;
+      }
+    });
+    // boolMap should be all-true (all conditions met)
+    const triggered = boolMap.every(entry => {
+      return entry === true;
+    });
+    context.flags['triggered'] = triggered;
+  },
+  'check-participant-registration': context => {
+    return null;
+  },
+  'create-report': context => {
+    // the output might define a profile
+    let profile = null;
+    if (context.action.output) {
+      const output = context.action.output[0];
+      if (output.profile) {
+        profile = output.profile[0];
+      }
+    }
+    let contentBundle = createBundle(context.records);
+    if (profile) {
+      contentBundle = validateProfile(contentBundle, profile);
+    }
+    const header = makeHeader();
+    context.contentBundle = contentBundle;
+    context.reportingBundle = createBundle([header, ...contentBundle.entry]);
+  },
+  'validate-report': context => {
+    // check if fhir is well-formed
+    const validation = fhir.validate(context.reportingBundle);
+    context.flags['valid'] = validation.valid;
+  },
+  'submit-report': context => {
+    context.client.dest.submit(context.reportingBundle);
+  },
+  'deidentify-report': context => {
+    const client = context.client.trustServices;
+    client.deidentify(context.reportingBundle).then(result => {
+      context.reportingBundle = result;
+    });
+    context.flags['deidentified'] = true;
```

Comment:
I think these trustService methods should be async. The deidentified flag will be set to true before the `context.reportingBundle` is updated. If there's no delay between this and the next action then its highly likely to be overwritten. 
This should also be coordinated with Dylan because his code is expecting these actions to run synchronously.

Reply:
The whole reporting workflow can be made async without too much trouble. I just made a quick change to mine that assumes each action can be async so it's called with await.